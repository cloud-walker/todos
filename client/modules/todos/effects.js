const ENDPOINT = 'http://localhost:3000/todos'

export const fetchTodos = () => fetch(ENDPOINT).then(res => res.json())

export const createTodo = data =>
  fetch(ENDPOINT, {
    method: 'POST',
    body: JSON.stringify(data),
    headers: {'Content-Type': 'application/json'},
  }).then(res => res.json())
