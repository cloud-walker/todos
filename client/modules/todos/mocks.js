export const todosStub = [
  {
    title: 'Comprare la nutella',
    id: 0,
  },
  {
    title: 'Pulire il bagno',
    id: 1,
  },
  {
    title: 'Preparare la valigia',
    id: 2,
  },
]

export const fetchTodosMock = () =>
  new Promise(resolve => {
    setTimeout(resolve, 250, todosStub)
  })
