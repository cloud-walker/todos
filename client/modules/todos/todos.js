import {useQuery, useMutation, queryCache} from 'react-query'

const ENDPOINT = 'http://localhost:3000/todos'

const fetchTodos = () => fetch(ENDPOINT).then(res => res.json())

const createTodo = data =>
  fetch(ENDPOINT, {
    method: 'POST',
    body: JSON.stringify(data),
    headers: {'Content-Type': 'application/json'},
  }).then(res => res.json())

export const useTodosQuery = () => useQuery(ENDPOINT, fetchTodos)

export const useTodoMutation = () =>
  useMutation(createTodo, {
    onSuccess: () => {
      queryCache.refetchQueries(ENDPOINT)
    },
  })
