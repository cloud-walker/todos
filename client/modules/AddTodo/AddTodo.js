import React from 'react'

import {useTodoMutation} from '../todos'

export const AddTodo = () => {
  const [value, setValue] = React.useState('')
  const [mutate] = useTodoMutation()
  const handleSubmit = e => {
    e.preventDefault()
    mutate({title: value})
    setValue('')
  }

  return (
    <form onSubmit={handleSubmit}>
      <label htmlFor="title" className="blue">
        Add your todo
      </label>
      <input
        type="text"
        id="title"
        value={value}
        onChange={e => setValue(e.target.value)}
      />

      <input type="submit" value="Add" />
    </form>
  )
}
