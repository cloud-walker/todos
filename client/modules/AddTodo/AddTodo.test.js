import {render, fireEvent, wait} from '@testing-library/react'
import React from 'react'

import {createTodo} from '../todos/effects'
import {AddTodo} from './AddTodo'

jest.mock('../todos/effects')

test('it should work properly', async () => {
  const createTodoMock = jest.fn(data => Promise.resolve({...data, id: 100}))
  createTodo.mockImplementation(createTodoMock)
  const {getByLabelText, getByText, container} = render(<AddTodo />)

  // Ci deve essere un input con la label "Add your todo"
  expect(getByLabelText('Add your todo')).toBeVisible()

  // Dobbiamo poter scrivere nell'input
  fireEvent.change(getByLabelText('Add your todo'), {
    target: {value: 'Comprare il pane'},
  })
  expect(getByLabelText('Add your todo')).toHaveValue('Comprare il pane')

  // Ci deve essere un tasto "Add"
  expect(getByText('Add')).toBeVisible()

  // Premendo il tasto "Add" viene lanciato il submit e viene resettato l'input
  fireEvent.click(getByText('Add'))

  await wait()

  expect(getByLabelText('Add your todo')).toHaveValue('')
  expect(createTodoMock).toHaveBeenCalled()
  expect(createTodoMock).toHaveBeenCalledWith({title: 'Comprare il pane'})
  expect(container.firstChild).toMatchSnapshot()
})
