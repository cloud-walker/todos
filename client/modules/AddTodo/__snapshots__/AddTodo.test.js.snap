// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should work properly 1`] = `
<form>
  <label
    class="blue"
    for="title"
  >
    Add your todo
  </label>
  <input
    id="title"
    type="text"
    value=""
  />
  <input
    type="submit"
    value="Add"
  />
</form>
`;
